openvasd/openvasd.c

```diff
@@ -0,0 +1,2329 @@
+  gchar *exclude_hosts;     /** String defining one or many hosts to exclude */
+  gchar *hosts;             /** String defining one or many hosts to scan */
+  gchar *ports;             /** String defining the ports to scan */
+  gchar *finished_hosts;    /** String defining hosts to exclude as finished */
+  gboolean icmp;            /** Alive test method icmp */
+  gboolean tcp_syn;         /** Alive test method tcp_syn */
+  gboolean tcp_ack;         /** Alive test method tcp_ack */
+  gboolean arp;             /** Alive test method arp */
+  gboolean consider_alive;  /** Alive test method consider alive */
+  int reverse_lookup_unify; /** Value defining reverse_lookup_unify opt */
+  int reverse_lookup_only;  /** Value defining reverse_lookup_only opt */
+};
+
+/**
+ * @brief Struct holding vt information
+ */
+struct openvasd_vt_single
+{
+  gchar *vt_id;
+  GHashTable *vt_values;
+};
+
+/**
+ * @brief Request methods
+ */
+enum openvas_request_method
+{
+  POST,
+  GET,
+  HEAD,
+  DELETE,
+};
+
+typedef enum openvas_request_method openvasd_req_method_t;
+
+/**
+ * @brief Allocate openvasd curl handler
+ *
+ * @return Openvasd curl handler.
+ */
+static openvasd_curlm_t *
+openvasd_curlm_handler_new (void)
+{
+  return (openvasd_curlm_t *) g_malloc0 (sizeof (struct openvasd_curlm));
+}
+
+/**
+ * @brief Cleanup an openvasd curl handler
+ *
+ * @param h Openvasd curl handler to clean
+ */
+static void
+openvasd_curlm_handler_close (openvasd_curlm_t **h)
+{
+  int queued = 0;
+
+  /* when an easy handle has completed, remove it */
+  CURLMsg *msg = curl_multi_info_read ((*h)->h, &queued);
+  if (msg)
+    {
+      if (msg->msg == CURLMSG_DONE)
+        {
+          curl_multi_remove_handle ((*h)->h, msg->easy_handle);
+          curl_easy_cleanup (msg->easy_handle);
+          curl_slist_free_all ((*h)->customheader);
+          curl_multi_cleanup ((*h)->h);
+          return;
+        }
+      g_warning ("%s: Not possible to clean up the curl handler", __func__);
+    }
+}
+
+/** @brief Allocate the vt stream struct to hold the response
+ *  and the curlm handler
+ *
+ *  @return The vt stream struct. Must be free with openvasd_vt_stream_new().
+ */
+static openvasd_vt_stream_t
+openvasd_vt_stream_new (void)
+{
+  openvasd_vt_stream_t s;
+  s = g_malloc0 (sizeof (struct openvasd_string));
+  s->len = 0;
+  s->ptr = g_malloc0 (s->len + 1);
+  s->curl_hnd = openvasd_curlm_handler_new ();
+  return s;
+}
+
+/** @brief Cleanup the string struct to hold the response and the
+ *  curl multiperform handler
+ *
+ *  @param s The string struct to be freed
+ */
+static void
+openvasd_vt_stream_cleanup (openvasd_vt_stream_t s)
+{
+  if (s)
+    g_free (s->ptr);
+
+  if (s->curl_hnd)
+    openvasd_curlm_handler_close (&(s->curl_hnd));
+}
+
+/** @brief Reinitialize the string struct to hold the response
+ *
+ *  @param s The string struct to be reset
+ */
+static void
+openvasd_vt_stream_reset (openvasd_vt_stream_t s)
```

Comment:
I get the feeling this function and `openvasd_reset_vt_stream` can be removed.

The stream will always be freed by `openvasd_connector_free`, and resetting it in gvmd is causing the stream memory to be freed in a loop, just to be allocated again (and also it's allocated tiny so it has to grow again through reallocation in `response_callback_fn`, from what I can tell).

Reply:
This function is only used when the curl multi handler is used. I use it to reset the string after processing a chunk of data but keeping the handler, which is used many times in this case. Otherwise, the string size will increased and kept until the feed updated finishes.